flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/threadinfo/ThreadInfoOperatorTracker.java

```diff
@@ -0,0 +1,407 @@
+                LOG.debug(
+                        "Triggering thread info sample for tasks: "
+                                + Arrays.toString(vertex.getTaskVertices()));
+            }
+
+            final AccessExecutionVertex[] executionVertices = vertex.getTaskVertices();
+            final CompletableFuture<ResourceManagerGateway> gatewayFuture =
+                    resourceManagerGatewayRetriever.getFuture();
+
+            CompletableFuture<OperatorThreadInfoStats> sample =
+                    gatewayFuture.thenCompose(
+                            (ResourceManagerGateway resourceManagerGateway) ->
+                                    coordinator.triggerThreadInfoRequest(
+                                            matchExecutionsWithGateways(
+                                                    executionVertices, resourceManagerGateway),
+                                            numSamples,
+                                            delayBetweenSamples,
+                                            maxThreadInfoDepth));
+
+            sample.whenCompleteAsync(new ThreadInfoSampleCompletionCallback(vertex), executor);
+        }
+    }
+
+    private List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
+            matchExecutionsWithGateways(
+                    AccessExecutionVertex[] executionVertices,
+                    ResourceManagerGateway resourceManagerGateway) {
+
+        List<Tuple2<AccessExecutionVertex, CompletableFuture<TaskExecutorGateway>>>
+                executionsWithGateways = new ArrayList<>();
+
+        for (AccessExecutionVertex executionVertex : executionVertices) {
+            TaskManagerLocation tmLocation = executionVertex.getCurrentAssignedResourceLocation();
+
+            if (tmLocation != null) {
+                CompletableFuture<TaskExecutorGateway> taskExecutorGatewayFuture =
+                        resourceManagerGateway.requestTaskExecutorGateway(
+                                tmLocation.getResourceID());
+
+                executionsWithGateways.add(
+                        new Tuple2<>(executionVertex, taskExecutorGatewayFuture));
+            } else {
+                LOG.warn("ExecutionVertex " + executionVertex + "is currently not assigned");
+            }
+        }
+
+        return executionsWithGateways;
+    }
+
+    @Override
+    public void cleanUpOperatorStatsCache() {
+        operatorStatsCache.cleanUp();
+    }
+
+    @Override
+    public void shutDown() {
+        synchronized (lock) {
+            if (!shutDown) {
+                operatorStatsCache.invalidateAll();
+                pendingStats.clear();
+
+                shutDown = true;
+            }
+        }
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> getResultAvailableFuture() {
+        return resultAvailableFuture;
+    }
+
+    /** Callback on completed thread info sample. */
+    class ThreadInfoSampleCompletionCallback
```

Comment:
why is this package-private?

Reply:
Mistakenly inherited from the old backpressure tracker. Moved to private.